docs/source/articles/implement-the-credit-card-tender-type.html.md

```diff
@@ -0,0 +1,1139 @@
+---
+title: Implement the Credit Card Tender Type
+excerpt: TODO
+created_at: 2019/11/05
+---
+
+Implement the Credit Card Tender Type
+======================================================================
+
+The credit card tender type is partially implemented, but has to be finished.
+
+Try to find a plugin that you can use, but if you can't, follow this process to finish the implementation on your own (or to create a plugin that you and others can use).
+
+
+Credit Card Gateway
+----------------------------------------------------------------------
+
+(_Explain_ gateways in the [Payment Tender Types](/articles/payment-tender-types.html) explain doc.)
+
+Credit card operation implementations (see [Payment Tender Types, Operation Implementations](/articles/payment-tender-types.html)) communicate with a gateway (see [Payment Tender Types, Gateway](/articles/payment-tender-types.html)).
+Within a credit card operation implementation, access to the gateway is provided by [`Payment::CreditCardOperation#gateway`](https://github.com/workarea-commerce/workarea/blob/master/core/app/models/workarea/payment/credit_card_operation.rb#L15-L17), which uses a gateway object that's created by an [initializer](https://github.com/workarea-commerce/workarea/blob/v3.4.18/core/config/initializers/11_payment.rb#L6).
+The gateway is an instance of [`ActiveMerchant::Billing::BogusGateway`](https://www.rubydoc.info/gems/activemerchant/1.99.0/ActiveMerchant/Billing/BogusGateway).
+
+That gateway is fine for local development and testing, but you'll want to use a different gateway for at least some uses cases.
+For example, in production you'll want to talk to a real credit card processing service over the network, so you need a gateway that does that.
+And in test, you'll want to talk to that same gateway, but use pre-recorded responses rather than going out to the real service.
+
+So you need to write your own initializer that has the logic for initializing the gateway and then inits the gateway according to that logic when the app boots.
+But you also need to override the gateway for some automated tests, and for that you need custom test support and test decorators.
+
+
+### Initializer
+
+Section introduction.
+
+
+#### Boilerplate
+
+```ruby
+# your_engine/config/initializers/credit_card_gateway.rb [1][2]
+
+credentials = Rails.application.secrets.credit_card_gateway #[2][3]
+
+if credentials.present? #[4]
+  Workarea.config.gateways.credit_card = #[5]
```

Comment:
Until the design of the credit card gateway extension point is changed in Workarea 3.6, I'm going to document the existing pattern of assigning a gateway instance to this config. We know this doesn't allow for different gateways for different sites in a multi site scenario, but we haven't encountered that being a problem for the credit card tender thus far.

I could write boilerplate that would work around this known issues in Workarea 3.5 and below, but there would be no concrete examples to show to the reader or for me to rely on to know that the solution works. I'd have to write my own solution and test it and then explain to the reader why the docs are different from what you see in actual plugins and application code.

But if anyone objects to my decision here, we can discuss it further.